File: 121.best-time-to-buy-and-sell-stock.rb
#
# @lc app=leetcode id=121 lang=ruby
#
# [121] Best Time to Buy and Sell Stock
#

# @lc code=start
# @param {Integer[]} prices
# @return {Integer}
def max_profit(prices)
  # At first, O(N^2) solution
  size = prices.size
  max_price = 0
  (0..size - 1).each do |start_index|
    (start_index + 1..size - 1).each do |end_index|
      if prices[end_index] - prices[start_index] > max_price
        max_price = prices[end_index] - prices[start_index]
      end
    end
  end

  max_price
end
# @lc code=end

File: 121.best-time-to-buy-and-sell-stock.go
/*
 * @lc app=leetcode id=121 lang=golang
 *
 * [121] Best Time to Buy and Sell Stock
 */
package main

// @lc code=start
func maxProfit(prices []int) int {
	minPrice := 100000
	maxProfit := 0

	for _, price := range prices {
		if price < minPrice {
			minPrice = price
			continue
		}

		profit := price - minPrice
		if profit > maxProfit {
			maxProfit = profit
		}
	}

	return maxProfit
}

// func maxProfit(prices []int) int {
// 	minPrice := 100000
// 	maxProfit := 0

// 	for i := 1; i <= len(prices); i++ {
// 		price := prices[i-1]
// 		if price < minPrice {
// 			minPrice = price
// 			continue
// 		}

// 		profit := price - minPrice
// 		if profit > maxProfit {
// 			maxProfit = profit
// 		}
// 	}

// 	return maxProfit
// }

// @lc code=end
